File: factory_method/FactoryMethod.java
package factory_method;

// product interface
interface Vehicle {
    void design();
    void manufacture();
}

// create concrete product classes
class Car implements Vehicle {
    @Override
    public void design() {
        System.out.println("Designing a Car");
    }

    @Override
    public void manufacture() {
        System.out.println("Manufacturing a Car");
    }
}

class Bike implements Vehicle {
    @Override
    public void design() {
        System.out.println("Designing a Bike");
    }

    @Override
    public void manufacture() {
        System.out.println("Manufacturing a Bike");
    }
}

// creator abstract class
abstract class VehicleFactory {
    abstract Vehicle createVehicle();

    public void buildVehicle() {
        Vehicle vehicle = createVehicle();
        vehicle.design();
        vehicle.manufacture();
    }
}

// concrete creators
class CarFactory extends VehicleFactory {
    @Override
    Vehicle createVehicle() {
        return new Car();
    }
}

class BikeFactory extends VehicleFactory {
    @Override
    Vehicle createVehicle() {
        return new Bike();
    }
}

public class FactoryMethod {
    public static void main(String[] args) {
        VehicleFactory carFactory = new CarFactory();
        carFactory.buildVehicle();

        VehicleFactory bikeFactory = new BikeFactory();
        bikeFactory.buildVehicle();
    }
}

File: factory_method/FactoryMethod.kt
package factory_method

// product interface
interface Vehicle {
    fun design()
    fun manufacture()
}

// create concrete product classes
class Car : Vehicle {
    override fun design() {
        println("Designing a Car")
    }

    override fun manufacture() {
        println("Manufacturing a Car")
    }
}

class Bike : Vehicle {
    override fun design() {
        println("Designing a Bike")
    }

    override fun manufacture() {
        println("Manufacturing a Bike")
    }
}

// creator abstract class
abstract class VehicleFactory {
    abstract fun createVehicle(): Vehicle

    fun buildVehicle() {
        val vehicle = createVehicle()
        vehicle.design()
        vehicle.manufacture()
    }
}

// concrete creators
class CarFactory : VehicleFactory() {
    override fun createVehicle(): Vehicle {
        return Car()
    }
}

class BikeFactory : VehicleFactory() {
    override fun createVehicle(): Vehicle {
        return Bike()
    }
}

fun main() {
    val carFactory: VehicleFactory = CarFactory()
    carFactory.buildVehicle()

    val bikeFactory: VehicleFactory = BikeFactory()
    bikeFactory.buildVehicle()
}